src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -255,8 +256,57 @@ private void RecordDependency(CallGraphNode callerKey, CallGraphNode calledKey,
         /// </summary>
         public List<ImmutableArray<CallGraphNode>> GetCallCycles()
         {
-            // ToDo: need to implement a robust algorithm for finding cycles
-            throw new NotImplementedException();
+            var callStack = new Dictionary<CallGraphNode, CallGraphNode>();
+            var finished = new HashSet<CallGraphNode>();
+            var cycles = new List<ImmutableArray<CallGraphNode>>();
+
+            void processDependencies(CallGraphNode node)
+            {
+                if (_Dependencies.TryGetValue(node, out var dependencies))
+                {
+                    foreach (var temp in dependencies)
+                    {
+                        var (curr, _) = temp;
+
+                        if (!finished.Contains(curr))
```

Comment:
Consider decreasing nesting:
```suggestion
                        if (finished.Contains(curr))
                        {
                            continue;
                        }
                        
                        // ...
```